### PR TITLE
Enforce retention policies

### PR DESCRIPTION
**Pull request by otoolep**

### CODE REVIEW

**Comment by otoolep** [comment]
This is work-in-progress, as I have some questions about this untested implementation, and also need to add unit-testing.

**Comment by otoolep** [comment]
To summarize the implementation follows. 

Each data node runs the "retention policy enforcement" loop, and looks for shard groups within each Retention Policy, which have an "end time" earlier than *now*. For all matching shard groups, a command is sent to the Raft cluster, requesting deletion of this shard group. Each data node, upon receipt of this command, deletes any shards it has associated with this shard group, and updates its local copy of the metastore.

Because all data nodes will run the loop, it does mean that multiple deletion commands may be generated for a given shard group -- the data nodes account for this by simply ignoring requests to delete non-existent shard groups.

File: server.go [hunk 1]
```diff
@@ -883,6 +930,69 @@ type createShardGroupIfNotExistsCommand struct {
...
+		shard.close()
+		if err := os.Remove(path); err != nil {
+			// Log, but keep going. This can happen if shards were deleted, but the server exited
+			// before it acknowledged the delete command.
+			log.Printf("error deleting shard %s, group ID %d, policy %s: %s", path, g.ID, rp.Name, err.Error())
+		}
+	}
+
+	// Remove from metastore.
+	rp.removeShardGroupByID(c.ID)
```
Comment (otoolep) [inline server.go]: It is these lines I have questions about for @benbjohnson around serlialization of database meta to metastore.

**Comment by otoolep** [comment]
Unit tests added, green build.

Still need to perform some system testing.

File: server.go (hunk 1 above)
Comment (otoolep) [inline server.go]: Answered by @benbjohnson and fixed by https://github.com/influxdb/influxdb/commit/beefdee425299d3e147142d0e956a81527aa14df

File: cmd/influxd/config.go [hunk 2]
```diff
@@ -92,14 +92,15 @@ type Config struct {
...
-		WriteBatchSize       int                       `toml:"write-batch-size"`
-		Engines              map[string]toml.Primitive `toml:"engines"`
-		RetentionSweepPeriod Duration                  `toml:"retention-sweep-period"`
+		Dir                   string                    `toml:"dir"`
+		Port                  int                       `toml:"port"`
+		WriteBufferSize       int                       `toml:"write-buffer-size"`
+		MaxOpenShards         int                       `toml:"max-open-shards"`
+		PointBatchSize        int                       `toml:"point-batch-size"`
+		WriteBatchSize        int                       `toml:"write-batch-size"`
+		Engines               map[string]toml.Primitive `toml:"engines"`
```
Comment (pauldix) [inline cmd/influxd/config.go]: Should just remove the `WriteBufferSize`, `MaxOpenShards`, `PointBatchSize`, `WriteBatchSize`, and `Engines` members on the `Data` struct

**Comment by pauldix** [comment]
One small comment, but other than that LGTM

**Comment by otoolep** [comment]
I performed a system test this morning, using an actual `influxd` process. Code looks good. I created a long (1 hour) and short (1 minute) retention policy, and wrote data to each. Within a couple of minutes (system had a retention check interval of 30 seconds) the short policy's shard groups were deleted. Log follows.

```
[srvr] 2015/02/11 10:18:04 broker enforcing retention policies with check interval of 30s
...
[srvr] 2015/02/11 10:18:34 retention policy enforcement check commencing
[srvr] 2015/02/11 10:19:04 retention policy enforcement check commencing
[srvr] 2015/02/11 10:19:34 retention policy enforcement check commencing
[srvr] 2015/02/11 10:20:04 retention policy enforcement check commencing
[srvr] 2015/02/11 10:20:04 shard group 1, retention policy short, database foo due for deletion
[srvr] 2015/02/11 10:20:34 retention policy enforcement check commencing
```
Queries of data in short returned no results, but queries of data in long did.

**Comment by otoolep** [comment]
This appears to be an integration test failure, I haven't seen it before. The integration tests may still be racy.

I'll kick Travis and try again.

**Comment by otoolep** [comment]
Green build, we appear to have some racy tests. This is currently captured by issue #1556.

File: server.go [hunk 3]
```diff
@@ -288,6 +294,47 @@ func (s *Server) load() error {
...
+// EnforceRetentionPolicies ensures that data that is aging-out due to retention policies
+// is removed from the server.
+func (s *Server) EnforceRetentionPolicies() {
+	log.Println("retention policy enforcement check commencing")
+
+	// Check all shard groups.
+	for _, db := range s.databases {
+		for _, rp := range db.policies {
+			for _, g := range rp.shardGroups {
+				if g.EndTime.Add(rp.Duration).Before(time.Now()) {
```
Comment (pauldix) [inline server.go]: Should be using UTC here, otherwise it's going to be off if their server isn't set to that.

**Pull request merged**